### PR TITLE
autoindex: trigger autoindex job from UI

### PR DESCRIPTION
**Pull request by Strum355**
Review recommended with: :ballot_box_with_check:Hide whitespace changes

Backend changes:
- Extracted the index queueing logic to a separate package to be reused in the graphql resolver and the frontend codeintel background indexing package
- Added a `ForceQueueIndex` to skip the logic that checks if an index is already queued or if an upload exists for the HEAD commit
- GraphQL mutation on the resolver to queue an index
- Extracted autoindexer metrics to separate package, changing to singleton with sync.Once

Frontend changes:
- `DynamicallyImportedMonacoSettingsEditor` (jsdoc to be added)
	- This can now take a custom `SaveToolbar`, if one needs additional functionality in the base `SaveToolbar`
	- As there is no inheritance in function components, only composition, and the custom `SaveToolbar` may have additional props ontop of the `SaveToolbar` props, a function that merges the `SaveToolbar` props with their custom additional props must be passed along with the custom toolbar `React.FunctionComponent`. This is defined/typified as `SaveToolbarPropsGenerator`
		- This fallsback to the original behaviour if nothing is set for the custom toolbar, and the generic T (which is the type of the custom toolbar props) defaults to {}

- `SaveToolbar`
	- `Props` became `React.PropsWithChildren<Props>` to allow custom toolbars to pass in additional react nodes to be included. Example can be seen in `CodeIntelAutoIndexSaveToolbar`, where we add an additional button and additional LoadingSpinner for the new state(s) that we can have
	- As custom toolbars may want to override when Save and Discard buttons are disabled and/or when errors are displayed, this is customizable by setting the `saveDiscardDisabled` and `willShowError` property functions. 
		- By default, the former was when `saving || !dirty` and the latter was `!saving`. Overriding the impl must also do this if they wish to append to existing functionality rather than change it.

- `CodeIntelAutoIndexSaveToolbar`
	- `React.FunctionComponent` that "implements" a toolbar by composition of a `SaveToolbar`. 
	- Overrides the save/discard button disabling and error displaying based on whether the "Queue Job" button was clicked

- `CodeIntelIndexConfigurationPage`
	- State is no longer binary saving|!saving, was expanded to Idle|Saving|Queueing
	- The `CodeIntelAutoIndexSaveToolbar` "Save", "Discard" and "Queue Job" buttons are disabled when either "Save" or "Queue Job" is clicked

**Failing tests to be addressed and additional in-code docs to be added**

--------

Nice Diagram : )
monakoeditor

### CODE REVIEW

File: client/web/src/components/SaveToolbar.tsx [hunk 1]
```diff
@@ -4,27 +4,43 @@ import CheckIcon from 'mdi-react/CheckIcon'
 import CloseIcon from 'mdi-react/CloseIcon'
 import * as React from 'react'
 
-interface Props {
+export interface Props {
```
Comment (eseliger) [inline client/web/src/components/SaveToolbar.tsx:7]: When it's exported, I'd make it a little more specific, because auto imports would be very bad otherwise

```suggestion
export interface SaveToolbarProps {
```

File: client/web/src/components/SaveToolbar.tsx [hunk 2]
```diff
@@ -33,22 +49,23 @@ export const SaveToolbar: React.FunctionComponent<Props> = ({ dirty, saving, err
...
                 <button
                     type="button"
-                    disabled={saveDiscardDisabled}
+                    disabled={disabled}
                     title={saveDiscardTitle || 'Save changes'}
                     className="btn btn-sm btn-success save-toolbar__item save-toolbar__btn save-toolbar__btn-first test-save-toolbar-save"
                     onClick={onSave}
                 >
-                    <CheckIcon className="icon-inline" /> Save changes
+                    <CheckIcon className="icon-inline" style={{ marginRight: '0.1em' }} /> Save changes
```
Comment (eseliger) [inline client/web/src/components/SaveToolbar.tsx:57]: inline styles are discouraged, why is that required? would "mr-1" or something work?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go [hunk 3]
```diff
@@ -155,6 +158,15 @@ func (r *Resolver) UpdateRepositoryIndexConfiguration(ctx context.Context, args
 	return &gql.EmptyResponse{}, nil
 }
 
+func (r *Resolver) QueueAutoIndexJobForRepo(ctx context.Context, id graphql.ID) (*gql.EmptyResponse, error) {
+	repositoryID, err := unmarshalLSIFIndexGQLID(id)
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go:162]: is that a repository or an lsif index? If it's an lsif index, is the name misleading in the graphql schema?

File: client/web/src/enterprise/codeintel/configuration/backend.ts [hunk 4]
```diff
@@ -65,3 +67,25 @@ export function updateConfiguration({ id, content }: { id: string; content: stri
...
+
+    return requestGraphQL<QueueAutoIndexJobForRepoResult, QueueAutoIndexJobForRepoVariables>(
+        query,
+        { id },
+    ).pipe(
+        map(dataOrThrowErrors),
+        map(data => {
+            if (!data.queueAutoIndexJobForRepo) {
+                throw createInvalidGraphQLMutationResponseError('QueueAutoIndexJobForRepo')
+            }
```
Comment (eseliger) [inline client/web/src/enterprise/codeintel/configuration/backend.ts:88]: Does that mean not returning an alwaysEmptyResolver has a meaning?

File: client/web/src/enterprise/codeintel/configuration/CodeIntelIndexConfigurationPage.tsx [hunk 5]
```diff
@@ -49,39 +58,77 @@ export const CodeIntelIndexConfigurationPage: FunctionComponent<CodeIntelIndexCo
...
+                    jsonSchema={allConfigSchema}
+                    canEdit={true}
+                    onSave={save}
+                    saving={saving}
+                    height={600}
+                    isLightTheme={isLightTheme}
+                    history={history}
+                    telemetryService={telemetryService}
+                    customSaveToolbar={customToolbar}
+                    onDirtyChange={(dirty: boolean) => setDirty(dirty)}
```
Comment (eseliger) [inline client/web/src/enterprise/codeintel/configuration/CodeIntelIndexConfigurationPage.tsx:130]: ideally, this would be done using 

```
const onDirtyChange = useCallback((dirty: boolean) => { setDirty(dirty) }, [])
```

to prevent rerenders each pass

File: client/web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx [hunk 6]
```diff
@@ -91,18 +99,33 @@ export class DynamicallyImportedMonacoSettingsEditor extends React.PureComponent
...
+                    (
+                        saveToolbar ||
+                        <SaveToolbar
+                            dirty={this.isDirty}
+                            saving={this.props.saving}
+                            onSave={this.onSave}
+                            onDiscard={this.discard}
+                        />
+                    )
+                }
```
Comment (eseliger) [inline client/web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx:128]: You could add an else branch above in l.113 so you dont need a second switch here :) but no blocker.
```suggestion
                        saveToolbar
                }
```

**Review by eseliger — APPROVED** [review]
Took a look mostly at frontend code :)

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go (hunk 3 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/resolver.go:162]: Oh wow yea youre right :facepalm:

File: client/web/src/components/SaveToolbar.tsx (hunk 2 above)
Comment (Strum355) [reply client/web/src/components/SaveToolbar.tsx:57]: The spacing between the icon and the text of the three buttons looked inconsistent with the Discard button, hence this. 

Without margin:
image

With margin:
image

File: client/web/src/enterprise/codeintel/configuration/backend.ts (hunk 4 above)
Comment (Strum355) [reply client/web/src/enterprise/codeintel/configuration/backend.ts:88]: Hmm Im not sure what you mean by this, I kinda just copied this from the function above :sweat_smile:

Comment (eseliger) [reply client/web/src/enterprise/codeintel/configuration/backend.ts:85]: I was just wondering if there are three states in the response

- Threw an error
- returned an empty response `{ queueAutoIndexJobForRepo: { alwaysNil: null } }`
- returned no empty response `{ queueAutoIndexJobForRepo: null }`

..because I think this check does this distinction 😬

Comment (efritz) [reply client/web/src/enterprise/codeintel/configuration/backend.ts:85]: Hmm I think that one could be simplified too to something like `map(data => {})`

Comment (eseliger) [reply client/web/src/enterprise/codeintel/configuration/backend.ts:85]: `mapTo(undefined)` should work then as well

File: enterprise/internal/codeintel/autoindex/enqueuer/enqueuer.go [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+func NewIndexEnqueuer(dbStore DBStore, gitClient GitserverClient, observationContext *observation.Context) *IndexEnqueuer {
+	return &IndexEnqueuer{
+		dbStore:         dbStore,
+		gitserverClient: gitClient,
+		operations:      observability.NewOperations(observationContext),
+	}
+}
+
+func (s *IndexEnqueuer) QueueIndex(ctx context.Context, repositoryID int) (err error) {
```
Comment (efritz) [inline enterprise/internal/codeintel/autoindex/enqueuer/enqueuer.go:34]: Can this _and_ `ForceQueueIndex` be written in terms of an internal method on this struct that does the same thin gbut only conditionally checks for `IsQueued`? Looks like that's the only difference.

File: client/web/src/components/SaveToolbar.tsx (hunk 2 above)
Comment (efritz) [reply client/web/src/components/SaveToolbar.tsx:68]: @Strum355 Could you post a ss with a `mr-1` class instead of the inline style for comparison?

**Review by efritz — APPROVED** [review]
Backend changes LGTM with a small comment that can be taken care of now or in a follow-up.

File: client/web/src/components/SaveToolbar.tsx (hunk 2 above)
Comment (Strum355) [reply client/web/src/components/SaveToolbar.tsx:68]: image
`mr-1` seems a bit too large a gap imo and is the smallest available option afaik

**Comment by Strum355** [comment]
@efritz fyi I extracted the indexer metrics stuff [here](https://github.com/sourcegraph/sourcegraph/pull/17101/commits/cd0de1954a89aadd416ab0b9cd55b39e4d237393) (pls ignore the tsx files, too much work to rebase those changes into a different commit)
